code/modules/atmospherics/machinery/components/turbine/turbine.dm

```diff
@@ -0,0 +1,372 @@
+	var/turf/horizontal = get_step(get_step(controller, turn(dir, 90)), turn(dir, 90))
+	var/turf/diagonal
+	if(dir & NORTH || dir & SOUTH)
+		diagonal = locate(horizontal.x, vertical.y, controller.z)
+	else
+		diagonal = locate(vertical.x, horizontal.y, controller.z)
+
+	for(var/turf/floor in block(vertical, diagonal))
+		for(var/obj/machinery/machine in floor.contents)
+			if(istype(machine, /obj/machinery/atmospherics/components/unary/turbine/turbine_inlet))
+				if(machine.dir == turn(dir, 270))
+					turbine_parts |= machine
+					inlet = machine
+					RegisterSignal(inlet, COMSIG_PARENT_QDELETING, .proc/disconnect_parts)
+			if(istype(machine, /obj/machinery/atmospherics/components/unary/turbine/turbine_outlet))
+				if(machine.dir == turn(dir, 90))
+					turbine_parts |= machine
+					outlet = machine
+					RegisterSignal(outlet, COMSIG_PARENT_QDELETING, .proc/disconnect_parts)
+			if(istype(machine, /obj/machinery/power/turbine/turbine_shaft))
+				if(machine.dir == turn(dir, 90))
+					turbine_parts |= machine
+					shaft = machine
+					RegisterSignal(shaft, COMSIG_PARENT_QDELETING, .proc/disconnect_parts)
+
+	if(turbine_parts.len == 3)
+		connected = TRUE
+		inlet.connected = TRUE
+		outlet.connected = TRUE
+		shaft.connected = TRUE
+		balloon_alert(user, "all parts connected")
+	else
+		if(!inlet)
+			balloon_alert(user, "turbine inlet missing or misplaced")
+		if(!outlet)
+			balloon_alert(user, "turbine outlet missing or misplaced")
+		if(!shaft)
+			balloon_alert(user, "turbine shaft missing or misplaced")
+
+	call_update_appearances()
+
+///Disconnect all parts, unregister signals
+/obj/machinery/power/turbine/turbine_controller/proc/disconnect_parts()
+	SIGNAL_HANDLER
+	connected = FALSE
+	if(inlet)
+		inlet.connected = FALSE
+		inlet = null
+	if(outlet)
+		outlet.connected = FALSE
+		outlet = null
+	if(shaft)
+		shaft.connected = FALSE
+		shaft = null
+	turbine_parts = list()
+	UnregisterSignal(inlet, COMSIG_PARENT_QDELETING)
+	UnregisterSignal(outlet, COMSIG_PARENT_QDELETING)
+	UnregisterSignal(shaft, COMSIG_PARENT_QDELETING)
+
+/obj/machinery/power/turbine/turbine_controller/process_atmos()
+	if(!connected || !on)
+		return
+
+	if(!inlet.airs[1].gases)
+		return
+
+	var/datum/gas_mixture/input_remove = inlet.airs[1].remove_ratio(input_ratio * 0.01)
+	var/datum/gas_mixture/output = outlet.airs[1]
+
+	if(!input_remove.heat_capacity())
+		return
+
+	heat_transfer_coefficient = 1
+	efficiency_coefficient = 0
+	rpm_coefficient = 1
+
+	check_gas_composition(input_remove)
+
+	first_point.merge(input_remove)
+	var/first_point_pressure = first_point.return_pressure()
+	var/first_point_temperature = first_point.return_temperature()
+	var/datum/gas_mixture/first_remove = first_point.remove(first_point.total_moles())
+
+	second_point.merge(first_remove)
+	var/second_point_pressure = second_point.return_pressure()
+	var/second_point_temperature = second_point.temperature
+	var/heat_capacity = second_point.heat_capacity()
+
```

Comment:
Doesn't this make second_point the same as first_point? or am I missing something

Reply:
the second point is similar to the first only for the temperature, tho i was thinking of allowing the gas to cool down when expanding since the second point has a bigger chamber than the first one

Reply:
i've made the second point to expand and lose a bit of temperature